modules/apigee_edge_teams/src/TeamPermissionHandler.php

```diff
@@ -0,0 +1,300 @@
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * version 2 as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
+ * MA 02110-1301, USA.
+ */
+
+namespace Drupal\apigee_edge_teams;
+
+use Drupal\apigee_edge_teams\Entity\TeamInterface;
+use Drupal\apigee_edge_teams\Entity\TeamRoleInterface;
+use Drupal\apigee_edge_teams\Exception\InvalidArgumentException;
+use Drupal\apigee_edge_teams\Structure\TeamPermission;
+use Drupal\Core\DependencyInjection\ClassResolverInterface;
+use Drupal\Core\Discovery\YamlDiscovery;
+use Drupal\Core\Entity\EntityTypeManagerInterface;
+use Drupal\Core\Extension\ModuleHandlerInterface;
+use Drupal\Core\Session\AccountInterface;
+use Drupal\Core\StringTranslation\StringTranslationTrait;
+
+/**
+ * Provides the available team permissions based on yml files.
+ *
+ * To define team permissions you can use a $module.team_permissions.yml file.
+ *
+ * If your module needs to define dynamic permissions you can use the
+ * permission_providers key to declare a class that implements
+ * DynamicTeamPermissionProviderInterface that will return an array of
+ * team permissions. Each item in the array can contain
+ * the same keys as an entry in $module.team_permissions.yml.
+ *
+ * Here is an example (comments have been added):
+ *
+ * @code
+ * # The key is the team permission machine name, and is required.
+ * manage team members:
+ *   # (required) The human-readable name of the team permission, to be shown
+ *   # on the team permission administration page.
+ *   title: 'Manage team members'
+ *   # (optional) Additional description for the team permission used in the UI.
+ *   description: 'Add/remove team members.'
+ *   # (optional) The category that the team permission belongs (ex.: Team Apps)
+ *   # to be shown on the team permission administration page.
+ *   # Default is the name of the provider module.
+ *   category: 'Members'
+ *
+ * # An array of classes used to generate dynamic team permissions.
+ * permission_providers:
+ *   # Each item in the array must implement
+ *   # DynamicTeamPermissionProviderInterface interface.
+ *   - Drupal\apigee_edge_teams\DefaultTeamPermissionsProvider
+ * @endcode
+ *
+ * Based on Drupal core's PermissionHandler with a bunch of improvements.
+ *
+ * @see \Drupal\user\PermissionHandler
+ * @see \Drupal\apigee_edge_teams\DynamicTeamPermissionProviderInterface
+ */
+final class TeamPermissionHandler implements TeamPermissionHandlerInterface {
```

Comment:
I wonder if this differs from a plugin manager so much that it can't extend `\Drupal\Component\Plugin\PluginManagerBase`.

Reply:
This could be plugin manager, but it could be an overkill for this, just like the user permission handler in Drupal 8 core, this is more lightweight and task-oriented but still allows to solve every _valid_ use cases that you may want to implement in a custom module. (Maybe it is even faster than if it would be a plugin manager.)

Let me know if you see any use case that you can not implement with this approach.